text/0756-helper-default-manager.md

```diff
@@ -0,0 +1,282 @@
+
+but there is a [TC39 proposal: proposal-deiter](https://github.com/tc39/proposal-deiter) that could make
+destructuring simpler and inlined to
+```ts
+  calculate(...numbers: number[], options: Options) {
+```
+
+#### Example with only positional parameters
+
+```hbs
+{{this.add 1 2 3 4}}
+```
+Because there are no named arguments passed in, the method signature can be simple:
+```js
+class A {
+  add(...numbers: number[]) {
+    // ...
+  }
+}
+```
+
+The implementation for the this function-handling helper-manager could look like this:
+```ts
+import {
+  setHelperManager,
+  capabilities as helperCapabilities,
+} from '@ember/helper';
+import { assert } from '@ember/debug';
+
+class FunctionHelperManager {
+  capabilities = helperCapabilities('3.23', {
+    hasValue: true,
+  });
+
+  createHelper(fn, args) {
+    return { fn, args };
+  }
+
+  getValue({ fn, args }) {
+    let argsForFn = args.positional;
+
+    if (Object.keys(args.named).length > 0) {
+      argsForFn.push(args.named);
+    } else {
+      argsForFn.push({});
+    }
+
+    return fn(...argsForFn);
+  }
+
+  getDebugName(fn) {
+    return fn.name || '(anonymous function)';
+  }
+}
+
+const DEFAULT_HELPER_MANAGER = new FunctionHelperManager();
+
+// side-effect -- this file needs to be imported for the helper manager to be installed
+setHelperManager(() => DEFAULT_HELPER_MANAGER, Function.prototype);
+```
+ - when the "helper" is created, the function is not invoked
+ - when `getValue` is invoked,
+   - the function is invoked with the named arguments all grouped into the last arg
+   - if no named arguments are given, an empty object is used instead to allow less nullish checking in userland
+ - to register this helper manager, it should occur during app boot so developers do not need to import anything to
+   trigger the `setHelperManager` call
+
+
+### How a template syntax plays in to this behavior
+
+In the current state of templates there is some ambiguity in syntax around helper/component invocation invocation.
+Below is a list exploring the various syntaxes and how the code implemented in the framework for this RFC
+will react to various passed value/function/etc types. All of this is current behavior and this RFC is not proposing
+a syntax change. In template strict mode, there is no ambiguity to worry about.
+
+- `{{val}}`
+  - `typeof val === 'function'`: Helper, invoked
+    - presently, it's possible to have curly components use this syntax as well. By defining this as a helper,
+      the possibility of using functions as components using curly syntax is eliminated. Curly components could still
```

Comment:
Hmm, I don't think this is correct: if you have an associated component manager with any object/function that component manager would be used instead of this default/fallback helper manager. I think this points out a missing piece of information in this RFC: the exact algorithm for choosing the manager to use. At a high level, I'd expect that to be:

* if inside element space
  * use `getModifierManager`
* if inside document body space using curly invocation
  1. attempt to lookup via `getComponentManager` (and use it)
  2. attempt to lookup via `getHelperManager` (and use it)
  3. if function -> fallback to this RFCs default manager
  4. render it ([Object object]?)
* if inside document body space using angle invocation
  1. attempt to lookup via `getComponentManager` (and use it)
* if inside of subexpression's "head" (e.g. `PathExpression`) position
  1. attempt to lookup via `getHelperManager` (and use it)
  2. if function -> fallback to this RFCs default manager
  3. error?
* if inside of a subexpression arguments
  1. pass the value